sklearn/metric_learning/nca.py

```diff
@@ -0,0 +1,280 @@
+from ..base import BaseEstimator, TransformerMixin
+from ..utils.validation import check_is_fitted
+
+import numpy as np
+import scipy as sp
+import scipy.misc
+import scipy.optimize
+import time
+
+import sys
+
+
+def nca_vectorized_oracle(X, y, n_components, loss, threshold=0.0):
+    dx = X[:, np.newaxis, :] - X[np.newaxis, :, :]  # n_samples x n_samples x n_features
+    outer = dx[:, :, np.newaxis, :] * dx[:, :, :, np.newaxis]  # n_samples x n_samples x n_features x n_features
```

Comment:
Another thing to check in addition to this is to unravel the `x_ik` to what they actually are: `x_ik[:, np.newaxis] * x_ik = (x_i - x_k)[:, np.newaxis] * (x_i - x_k) = x_i[:, np.newaxis] * x_i + ...` which yields 4 terms, all of which you can already find in your `dx`. This should bring down memory consumption.